File: 6.Basic-Sorting/Selection/Example/Selection.java
/**
 * Implements the Selection Sort algorithm to sort an array in ascending order.
 *
 * @param {int[]} arr - The array of numbers to be sorted.
 * @return {int[]} - The sorted array.
 *
 * @timeComplexity O(n^2) - Worst-case, average-case, and best-case.
 *                  The number of comparisons and swaps grows quadratically with the input size (n).
 * @spaceComplexity O(1) - In-place sorting, modifies the original array without additional data structures.
 *
 * @additionalNotes
 *  - Sorting and Heapselect: Selection Sort is a comparison-based sorting algorithm, unlike Heap Sort
 *    which utilizes a heap data structure for efficient sorting (O(n log n) average-case). While Selection Sort
 *    is simpler, Heap Sort generally has better performance.
 *  - Pivoting: Selection Sort does not use pivoting. Pivoting is a key concept in algorithms like Quick Sort
 *    that divide and conquer the input array by partitioning it based on a chosen pivot element. Selection Sort
 *    progressively finds the minimum element within the unsorted portion and swaps it into place.
 *  - Factories: Selection Sort can be implemented using a design pattern called a factory. A factory function
 *    could create sorting function instances configured for different data types or comparison logic.
 *  - Parallel Algorithms: Selection Sort is not highly parallelizable due to its inherent dependencies between
 *    comparisons and swaps. Other sorting algorithms like Merge Sort are more suitable for parallel execution
 *    on multi-core processors.
 */
public class SelectionSort {

    public static int[] sort(int[] arr) {
    // Loop through each element of the array (excluding the last element)
    for (int i = 0; i < arr.length - 1; i++) {
      int minIndex = i;

      // Find the index of the minimum element in the unsorted part
      for (int j = i + 1; j < arr.length; j++) {
        if (arr[j] < arr[minIndex]) {
          minIndex = j;
        }
      }

      // Swap the element at the current index with the minimum element if necessary
      if (minIndex != i) {
        int temp = arr[i];
        arr[i] = arr[minIndex];
        arr[minIndex] = temp;
      }
    }

    return arr;
  }

    public static void main(String[] args) {
    int[] unsortedArray = {64, 25, 12, 22, 11};
    int[] sortedArray = sort(unsortedArray);

    System.out.println("Original Array: " + java.util.Arrays.toString(unsortedArray));
    System.out.println("Sorted Array : " + java.util.Arrays.toString(sortedArray));
  }
}

File: 1.Elementary/6.Basic-Sorting/Selection/Example/Selection.py
"""
Implements the Selection Sort algorithm to sort a list in ascending order.

This function iterates through the list, finding the minimum element
within the unsorted portion and swapping it with the element at the
current position. This process repeats for each element in the list.

Args:
  arr: The list of numbers to be sorted.

Returns:
  The sorted list.

Time Complexity: O(n^2) - Worst-case, average-case, and best-case.
                  The number of comparisons and swaps grows quadratically with the input size (n).
Space Complexity: O(1) - In-place sorting, modifies the original list without additional data structures.
"""
def selection_sort(arr):
  # Loop through each element of the list (excluding the last element)
  for i in range(len(arr) - 1):
    min_index = i

    # Find the index of the minimum element in the unsorted part
    for j in range(i + 1, len(arr)):
      if arr[j] < arr[min_index]:
        min_index = j

    # Swap the element at the current index with the minimum element if necessary
    if min_index != i:
      arr[i], arr[min_index] = arr[min_index], arr[i]

  return arr

# Example Usage
unsorted_list = [64, 25, 12, 22, 11]
sorted_list = selection_sort(unsorted_list.copy())  # Sort a copy to avoid modifying original list
print("Original List:", unsorted_list)  # Output: [64, 25, 12, 22, 11]
print("Sorted List :", sorted_list)     # Output: [11, 12, 22, 25, 64]
